doc/computing/randomness.rst

```diff
@@ -0,0 +1,315 @@
+    >>> a = RandomForestClassifier(random_state=rng)
+    >>> b = clone(a)
+
+Since a `RandomState` instance was passed to `a`, `a` and `b` are not clones
+in the strict sense, but rather clones in the statistical sense: `a` and `b`
+will still be different models, even after calling `fit(X, y)` on the same
+data. Moreover, `a` and `b` will influence each-other since they share the
+same internal RNG: calling `a.fit` will consume `b`'s RNG, and calling
+`b.fit` will consume `a`'s RNG, since they are the same. This is true for
+any estimators that share a `random_state` parameter; it is not specific to
+clones.
+
+If an int were passed, `a` and `b` would be exact clones and they would not
+influence each other.
+
+This is an important thing to remember because :func:`~sklearn.clone` is
+called everywhere in scikit-learn tools: in particular, most meta-estimators
+that accept non-fitted estimators will in fact call :func:`~sklearn.clone`
+internally (:class:`~sklearn.model_selection.GridSearchCV`,
+:class:`~sklearn.ensemble.StackingClassifier`,
+:class:`~sklearn.calibration.CalibratedClassifierCV`, etc.).
+
+CV splitters
+............
+
+When passed a `RandomState` instance, cv splitters yield different splits
+each time `split` is called. This can lead to dramatic mistakes when
+comparing the performance of different estimators::
+
+    >>> from sklearn.ensemble import RandomForestClassifier
+    >>> from sklearn.ensemble import GradientBoostingClassifier
+    >>> from sklearn.datasets import make_classification
+    >>> from sklearn.model_selection import KFold
+    >>> from sklearn.model_selection import cross_val_score
+    >>> import numpy as np
+
+    >>> rng = np.random.RandomState(0)
+    >>> X, y = make_classification(random_state=rng)
+    >>> rf = RandomForestClassifier(random_state=rng)
+    >>> gbdt = GradientBoostingClassifier(random_state=rng)
+    >>> cv = KFold(shuffle=True, random_state=rng)
+
+    >>> for est in (rf, gbdt):
+    ...     print(cross_val_score(est, X, y, cv=cv))
+    [0.85 0.95 0.9  0.95 0.95]
+    [0.85 0.7  0.95 0.8  0.85]
+
+Directly comparing the performance of the random forest vs the gradient
+boosting estimator on each fold would be a methodological mistake: **the
+splits on which the estimators are evaluated are different**. Indeed,
+:func:`~sklearn.model_selection.cross_val_score` will internally call
+`cv.split` on the same :class:`~sklearn.model_selection.KFold` instance,
+but the splits will be different each time. This is also true for any tool
+that performs model selection via cross-validation, e.g.
+:class:`~sklearn.model_selection.GridSearchCV` and
+:class:`~sklearn.model_selection.RandomizedSearchCV`: scores are not
+comparable fold-to-fold across different calls to `search.fit`, since
+`cv.split` would have been called multiple times. Within a single call to
+`search.fit`, however, fold-to-fold comparison is possible since the search
+estimator only calls `cv.split` once.
```

Comment:
This paragraph lacks nuance: the estimate of the performance of the two estimators are still unbiased, they just have an uncontrolled variance. The error would be to substract one list to the other.

Reply:
I'm not sure what to do here: bias isn't the topic of this paragraph. We discuss these things below I believe.

Reply:
To be more direct: it is not a mistake to compare the performance of the two models, because this comparison is on expectation correct. The paragraph above implies that it is.

What would be a mistake would be to perform a paired test, in other words comparing each entry of one list to the corresponding entry of the other. However the only difference between comparing entries one to one and comparing the mean is the variance, and the problem of estimating the variance of the error from a cross-validation procedure is an unsolved one.

Reply:
I think this paragraph was updated since the time you first commented on it. Specifically, it's now clear IMHO that what we say is that:

- fold-to-fold comparison is a mistake
- comparing average scores is OK as long as there are enough folds and data.

Could you double check?

Reply:
OK, I missed the "on each fold" above. Maybe we should put it in bold.